File: src/recipe.test.ts
import { describe, it, expect } from "bun:test";
import { parseRecipe, type Recipe } from "./recipe";

describe("parseRecipe", () => {
  it("parses a valid recipe", () => {
    const yaml = `
name: cve-review
prompt: "Review CVEs for staleness"
scope:
  domain: [security]
  entities: [acme]
  search: ["patch"]
  path: worlds/security/
auto_merge: false
steps:
  - mode: prune
    model: gemini-2.0-flash
    prompt: "Find stale CVEs"
  - mode: distill
    model: claude-sonnet-4-6
    prompt: "Find patterns"
`;
    const recipe = parseRecipe(yaml);
    expect(recipe.name).toBe("cve-review");
    expect(recipe.steps).toHaveLength(2);
    expect(recipe.steps[0].mode).toBe("prune");
    expect(recipe.steps[0].model).toBe("gemini-2.0-flash");
    expect(recipe.steps[1].mode).toBe("distill");
    expect(recipe.scope.domain).toEqual(["security"]);
    expect(recipe.scope.entities).toEqual(["acme"]);
    expect(recipe.scope.search).toEqual(["patch"]);
    expect(recipe.scope.path).toBe("worlds/security/");
    expect(recipe.auto_merge).toBe(false);
  });

  it("defaults auto_merge to false", () => {
    const yaml = `
name: test
prompt: ""
steps:
  - mode: prune
`;
    const recipe = parseRecipe(yaml);
    expect(recipe.auto_merge).toBe(false);
  });

  it("throws on missing name", () => {
    const yaml = `
steps:
  - mode: prune
`;
    expect(() => parseRecipe(yaml)).toThrow();
  });

  it("throws on invalid mode", () => {
    const yaml = `
name: test
steps:
  - mode: invalid
`;
    expect(() => parseRecipe(yaml)).toThrow();
  });

  it("defaults empty scope fields", () => {
    const yaml = `
name: test
scope:
  domain: []
steps:
  - mode: prune
`;
    const recipe = parseRecipe(yaml);
    expect(recipe.scope!.domain).toEqual([]);
    expect(recipe.scope!.entities).toEqual([]);
    expect(recipe.scope!.search).toEqual([]);
    expect(recipe.scope!.path).toBe("");
  });

  it("allows omitting scope entirely for auto-discovery", () => {
    const yaml = `
name: auto-test
steps:
  - mode: prune
`;
    const recipe = parseRecipe(yaml);
    expect(recipe.scope).toBeUndefined();
  });
});

File: internal/synthesize/recipe_test.go
package synthesize

import (
	"testing"
)

const testRecipeYAML = `
name: my-recipe
prompt: "summarize the knowledge base"
auto_merge: true
steps:
  - mode: prune
    model: claude-sonnet-4-6
  - mode: distill
    max_depth: 2
    resolution: 1.5
`

func TestParseRecipe(t *testing.T) {
	r, err := ParseRecipe(testRecipeYAML)
	if err != nil {
		t.Fatalf("ParseRecipe returned error: %v", err)
	}

	if r.Name != "my-recipe" {
		t.Errorf("Name: got %q, want %q", r.Name, "my-recipe")
	}
	if r.Prompt != "summarize the knowledge base" {
		t.Errorf("Prompt: got %q, want %q", r.Prompt, "summarize the knowledge base")
	}
	if !r.AutoMerge {
		t.Error("AutoMerge: got false, want true")
	}
	if r.Scope != nil {
		t.Errorf("Scope: got %+v, want nil (auto-discovery)", r.Scope)
	}
	if len(r.Steps) != 2 {
		t.Fatalf("Steps: got %d, want 2", len(r.Steps))
	}

	step0 := r.Steps[0]
	if step0.Mode != "prune" {
		t.Errorf("Steps[0].Mode: got %q, want %q", step0.Mode, "prune")
	}
	if step0.Model != "claude-sonnet-4-6" {
		t.Errorf("Steps[0].Model: got %q, want %q", step0.Model, "claude-sonnet-4-6")
	}

	step1 := r.Steps[1]
	if step1.Mode != "distill" {
		t.Errorf("Steps[1].Mode: got %q, want %q", step1.Mode, "distill")
	}
	if step1.MaxDepth != 2 {
		t.Errorf("Steps[1].MaxDepth: got %d, want 2", step1.MaxDepth)
	}
	if step1.Resolution != 1.5 {
		t.Errorf("Steps[1].Resolution: got %f, want 1.5", step1.Resolution)
	}
}

func TestParseRecipeWithProfile(t *testing.T) {
	yml := `
name: profiled
steps:
  - mode: prune
    profile: small
    retry_on_passive: false
  - mode: distill
`
	r, err := ParseRecipe(yml)
	if err != nil {
		t.Fatalf("ParseRecipe: %v", err)
	}
	if r.Steps[0].Profile != "small" {
		t.Errorf("Steps[0].Profile: got %q, want %q", r.Steps[0].Profile, "small")
	}
	if r.Steps[0].RetryOnPassive == nil || *r.Steps[0].RetryOnPassive != false {
		t.Errorf("Steps[0].RetryOnPassive: want pointer to false")
	}
	if r.Steps[1].Profile != "" {
		t.Errorf("Steps[1].Profile: got %q, want empty", r.Steps[1].Profile)
	}
	if r.Steps[1].RetryOnPassive != nil {
		t.Errorf("Steps[1].RetryOnPassive: want nil")
	}
}

func TestParseRecipeValidation(t *testing.T) {
	tests := []struct {
		name    string
		yaml    string
		wantErr bool
	}{
		{
			name:    "missing name",
			yaml:    "steps:\n  - mode: prune\n",
			wantErr: true,
		},
		{
			name:    "no steps",
			yaml:    "name: foo\n",
			wantErr: true,
		},
		{
			name:    "invalid mode",
			yaml:    "name: foo\nsteps:\n  - mode: invalid\n",
			wantErr: true,
		},
		{
			name: "valid with scope",
			yaml: `name: scoped
steps:
  - mode: prune
scope:
  domain: [testing]
  path: know/test/
`,
			wantErr: false,
		},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			_, err := ParseRecipe(tc.yaml)
			if (err != nil) != tc.wantErr {
				t.Errorf("ParseRecipe error=%v, wantErr=%v", err, tc.wantErr)
			}
		})
	}
}
